File: ui-tests/PageObjects/Panes/Terminal/TerminalLine.cs
using System;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Playwright;

namespace UiTests.PageObjects.Panes.Terminal;

/// <summary>
/// Represents a single rendered line inside the terminal output pane.
/// </summary>
public class TerminalLine
{
    private readonly ILocator _root;

    public TerminalLine(ILocator root)
    {
        _root = root ?? throw new ArgumentNullException(nameof(root));
    }

    /// <summary>
    /// Locator of the terminal line container.
    /// </summary>
    public ILocator Root => _root;

    private ILocator StateContainer()
        => _root.Locator("> div").First;

    /// <summary>
    /// Zero-based identifier extracted from the line DOM id.
    /// </summary>
    public async Task<int> LineIndexAsync()
    {
        var idAttr = await _root.GetAttributeAsync("id");
        if (idAttr is null)
        {
            return -1;
        }

        var suffix = idAttr.Split('-').LastOrDefault();
        return int.TryParse(suffix, out var index) ? index : -1;
    }

    /// <summary>
    /// Returns the current temporal state (past, active, future) inferred from CSS classes.
    /// </summary>
    public async Task<string> StateAsync()
    {
        var classAttr = await StateContainer().GetAttributeAsync("class") ?? string.Empty;
        return classAttr.Split(' ', StringSplitOptions.RemoveEmptyEntries).FirstOrDefault() ?? string.Empty;
    }

    /// <summary>
    /// Indicates whether the line is styled as grayed-out (future output).
    /// </summary>
    public async Task<bool> IsGrayedOutAsync()
    {
        var state = await StateAsync();
        return string.Equals(state, "future", StringComparison.OrdinalIgnoreCase);
    }

    /// <summary>
    /// Extracts the textual content of the line.
    /// </summary>
    public async Task<string> TextAsync()
    {
        var text = await _root.InnerTextAsync();
        return text?.Trim() ?? string.Empty;
    }

    /// <summary>
    /// Clicks the terminal line using the specified mouse button.
    /// </summary>
    public Task ClickAsync(MouseButton button = MouseButton.Left)
        => _root.ClickAsync(new() { Button = button });
}

File: tsc-ui-tests/page-objects/panes/terminal/terminal-line.ts
import type { Locator } from "@playwright/test";

/**
 * Represents a single rendered line inside the terminal output pane.
 *
 * Port of ui-tests/PageObjects/Panes/Terminal/TerminalLine.cs
 */
export class TerminalLine {
  readonly root: Locator;

  constructor(root: Locator) {
    this.root = root;
  }

  private stateContainer(): Locator {
    return this.root.locator("> div").first();
  }

  async lineIndex(): Promise<number> {
    const idAttr = await this.root.getAttribute("id");
    if (!idAttr) return -1;
    const suffix = idAttr.split("-").pop();
    return suffix ? parseInt(suffix, 10) || -1 : -1;
  }

  /**
   * Returns the current temporal state (past, active, future) from CSS classes.
   */
  async state(): Promise<string> {
    const classAttr = (await this.stateContainer().getAttribute("class")) ?? "";
    return classAttr.split(/\s+/).filter(Boolean)[0] ?? "";
  }

  async isGrayedOut(): Promise<boolean> {
    const state = await this.state();
    return state.toLowerCase() === "future";
  }

  async text(): Promise<string> {
    return (await this.root.innerText()).trim();
  }

  async click(button: "left" | "right" = "left"): Promise<void> {
    await this.root.click({ button });
  }
}
